### DISCUSSION TITLE
What's the difference between `useHead` and `useServerSeoMeta`

### DISCUSSION THREAD

**Discussion by ghost** [nuxt/nuxt#18554] [Questions]
I am a Nuxt novice and I noticed that both `useHead` and `useServerSeoMeta` can set `<title>` content, but I am not sure about the difference between them, would you mind explaining it or providing more details in the [documentation](https://nuxt.com/docs/getting-started/seo-meta)?

```js
useHead({
  title: 'Nuxt',
})

useServerSeoMeta({
  title: 'Nuxt Again',
})
```

Thanks! ❤️

**Comment by ndragun92** [comment]
In most instances, the meta does not need to be reactive as robots will only scan the initial load. So we recommend using `useServerSeoMeta` as a performance-focused utility that will not do anything (or return a `head` object) on the client.

It is advised to use `useServerSeoMeta` unless you _need_ client-side rendering of your SEO meta tags.

The `useServerSeoMeta` composable is identical to `useSeoMeta` except that it will have no effect (and will return nothing) if called on the client.

Hope it helps

**Reply by ghost** [reply]
Thank you for your informative reply. I understand the difference between `useSeoMeta` and `useServerSeoMeta` now. However, I am still curious about the difference between `useSeoMeta` and `useHead`. Could you please explain or provide more information? Thank you in advance.

**Reply by ndragun92** [reply]
## Composable: `useSeoMeta`

The `useSeoMeta` composable lets you define your sites SEO Meta as a flat object with full TypeScript support.

This helps you avoid common mistakes, such as using `name` instead of `property` as well typos.

### Example

#### Simple

```vue{}[app.vue]
<script setup lang="ts">
useSeoMeta({
  title: 'My Amazing Site',
  ogTitle: 'My Amazing Site',
  description: 'This is my amazing site, let me tell you all about it.',
  ogDescription: 'This is my amazing site, let me tell you all about it.',
  ogImage: 'https://example.com/image.png',
  twitterCard: 'summary_large_image',
})
</script>
```

#### Reactive

When inserting tags that are reactive, for example, from an API request, you should
use the computed getter syntax, the same as `useHead`.

```vue{}[app.vue]
<script setup lang="ts">
const data = useFetch(() => $fetch('/api/example'))
useSeoMeta({
  ogTitle: () => `${data.value?.title} - My Site`,
  description: () => data.value?.description,
  ogDescription: () => data.value?.description,
})
</script>
```

### Client-side optimisation

In most instances, the meta does not need to be reactive as robots will only scan the initial load.

The composable itself is a ~2kB, so you may consider only using it on the server and having it be treeshaken from the client bundle.

```vue{}[app.vue]
<script setup lang="ts">
// only run on the server or client side dev
if (process.dev || process.server) {
  useSeoMeta({ description: () => myDescription.value })
}
</script>
```

Read More: https://unhead.unjs.io/usage/composables/use-seo-meta#useserverseometa



So in short I think useSeoMeta was made especially for Nuxt and SSR support no that major difference I think

**Reply by ghost** [reply]
Thank you for your response, it has cleared up some of my doubts. However, I still have further questions and would appreciate more information.

**Reply by Naybnet** [reply]
Thank you @ndragun92, your answer is also super helpful to me, and the mention of the treeshaking is very valuable, I had not thought about it before.

**Reply by qtipee** [reply]
Hello :)

Just to be sure, if we use `useServerSeoMeta` so that it is server side only, we should use `useHead` to update the tab title client side?

**Reply by riddla** [reply]
@qtipee, client side updates might also need using the composable in a reactive way, see https://unhead.harlanzw.com/integrations/vue/reactivity#reactivity.

(Depending on where you get your content for the meta tags etc. from. In my case I used data defined with `definePageMeta` and was wondering why the `<title/>` was not updating on client-side navigation.)

**Reply by gaomingzhao666** [reply]
Thank you so much for this detailed and helpful description (●'◡'●)

**Reply by MickL** [reply]
Hey guys! I dont fully understand:

- You said `useServerSeoMeta()` only runs on client side
- If I use `useServerSeoMeta()` to set the title, the title changes when I navigate on client side so this doesnt seem to be fully correct?
- Also when using `useServerSeoMeta()` and setting `description`, it gets overridden by whatever is set in nuxt.config `site.description`

**Reply by gaomingzhao666** [reply]
> Hey guys! I dont fully understand:
> 
> * You said `useServerSeoMeta()` only runs on client side
> * If I use `useServerSeoMeta()` to set the title, the title changes when I navigate on client side so this doesnt seem to be fully correct?
> * Also when using `useServerSeoMeta()` and setting `description`, it gets overridden by whatever is set in nuxt.config `site.description`

Hello, refer to https://github.com/nuxt-modules/i18n/discussions/2713, hope this can help you

**Reply by MickL** [reply]
No this doesnt really answer my questions

**Reply by gaomingzhao666** [reply]
@MickL In my opinion.
> * If I use `useServerSeoMeta()` to set the title, the title changes when I navigate on client side so this doesnt seem to be fully correct?

No, it's still correct, and the title changes when router changes, because composable `useServerSeoMeta` is described as unreactive, meta data was defined in server-side, so the meta data will changes when page (router) changed.

> * Also when using `useServerSeoMeta()` and setting `description`, it gets overridden by whatever is set in nuxt.config `site.description`

No, the site.description is used for the description of entire website, and the useServerSeoMeta's description is used for specific page.

**Reply by MickL** [reply]
My last point might be a bug of nuxt-seo, issue: https://github.com/harlan-zw/nuxt-seo/issues/389

What I dont understand is why `useServerSeoMeta()` is changing the title (which is a good thing!) when is doesnt run on client. Minimal reproduction repo: https://stackblitz.com/edit/nuxt-starter-tcq8ttrk?file=pages%2Findex.vue,pages%2Fabout.vue

I feel like this needs better clarification in the docs.

**Reply by okainov** [reply]
Now it's second  half of 2025, but the question somehow still stands... Now there is only `useHead` and `useSeoMeta` (other ones were deprecated in favor of it)

Documentation page clearly officially describes how to make tags reactive https://nuxt.com/docs/4.x/api/composables/use-seo-meta:

> When inserting tags that are reactive, you should use the computed getter syntax (`() => value`):

I might be dumb, but I stuggle to understand why would one want e.g. `title` to be NOT reactive. When I navigate the pages on the client as in SPA, I do want to see updated title for each new page, right? And that's what is clearly told by [Performance](https://nuxt.com/docs/4.x/api/composables/use-seo-meta#performance) part of the docs.

I think what I miss from the docs is some clear recommendation like:

> What should I use?
> In most cases, you should use `useSeoMeta`, unless you're adding some extra tags to `head` not covered by it - use `useHead` in this case.
> Only use `definePageMeta` for actual internal page properties, such as `layout` or `middleware`, but not for title.

And all the places in the docs which set page title via `definePageMeta` should be updated to use `useSeoMeta`.

Does this make sense?
